PayloadPkg/OsLoader/PreOsChecker.c

```diff
@@ -0,0 +1,152 @@
+/** @file
+
+  Copyright (c) 2019, Intel Corporation. All rights reserved.<BR>
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#include <Library/DebugLib.h>
+#include <Library/ContainerLib.h>
+#include <Library/ElfLib.h>
+#include <Library/LinuxLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/BootloaderCommonLib.h>
+
+#define SIGNATURE_IPFW    SIGNATURE_32 ('I', 'P', 'F', 'W')
+#define SIGNATURE_POSC    SIGNATURE_32 ('P', 'O', 'S', 'C')
+
+typedef struct {
+  UINT32 Eax;
+  UINT32 Ebx;
+  UINT32 Ecx;
+  UINT32 Edx;
+  UINT32 Esi;
+  UINT32 Edi;
+  UINT32 Ebp;
+  UINT32 Eip;
+  UINT32 Eflags;
+} OS_BOOT_STATE;
+
+typedef struct {
+  UINT32        Version;
+  UINT32        HeapSize;
+  UINT32        HeapAddr;
+  OS_BOOT_STATE OsBootState;
+  UINT32        HobListPtr;
+} PRE_OS_PAYLOAD_PARAM;
+
+typedef VOID (*PRE_OS_CHECKER_ENTRY) (VOID *Params);
+
+STATIC VOID *mPreOsCheckerEntry = NULL;
+
+/**
+  To see whether PreOsChecker is loaded properly
+
+  @retval           TRUE if successfully loaded, otherwise FALSE
+
+**/
+BOOLEAN
+EFIAPI
+IsPreOsCheckerLoaded (
+  IN  VOID
+  )
+{
+  return ((mPreOsCheckerEntry == NULL) ? FALSE : TRUE);
+}
+
+/**
+  Search for and load pre-OS checker for execution.
+
+  This function will search for a pre-OS checker binary that might
+  be loaded as part of the payload, if found get the entry point for
+  execution later instead of jumping into the OS directly.
+
+  @retval  EFI_LOAD_ERROR         Pre-OS checker is not loaded successfully.
+  @retval  EFI_INVALID_PARAMETER  Invalid parameter is used at loading PreOsChecker image
+  @retval  EFI_UNSUPPORTED        Pre-OS checker image is not supported
+
+**/
+EFI_STATUS
+EFIAPI
+LoadPreOsChecker (
+  IN  VOID
+  )
+{
+  EFI_STATUS         Status;
+  UINT32             Length;
+  VOID              *Buffer;
+
+  //
+  // TBD: Set to NULL to re-load PreOsChecker in case of restarting Payload for now.
+  //
+  mPreOsCheckerEntry = NULL;
+  //if (mPreOsCheckerEntry != NULL) {
+  //  DEBUG ((DEBUG_INFO, "Pre-OS checker is already loaded at entry @ 0x%p\n", mPreOsCheckerEntry));
+  //  return EFI_STATUS;
+  //}
+
+  //
+  // Check if there is a pre-OS checker that needs to be executed
+  //
+  Buffer = NULL;
+  Length = 0;
+  Status = LoadComponent (SIGNATURE_IPFW, SIGNATURE_POSC, &Buffer, &Length);
```

Comment:
Do we have a way to indicate if PreOsChecker support is required or not ?   
If a platform does not need PreOsChecker,  maybe we should not try to load the image at all.

Reply:
Updated in coming push. Token space has been updated. Will follow up on-going boards which will be impacted by this change once this is merged.